src/java/com/zimbra/oauth/handlers/impl/LinkedinOAuth2Handler.java

```diff
@@ -0,0 +1,231 @@
+            constant = value;
+        }
+
+        public String getValue() {
+            return constant;
+        }
+
+        public static LinkedinErrorCodes fromString(String value) {
+            for (LinkedinErrorCodes code : LinkedinErrorCodes.values()) {
+                if (code.getValue().equals(value)) {
+                    return code;
+                }
+            }
+            return LinkedinErrorCodes.DEFAULT_ERROR;
+        }
+    }
+
+    public LinkedinOAuth2Handler(Configuration config) {
+        super(config, LinkedinOAuth2Constants.CLIENT_NAME.getValue(), LinkedinOAuth2Constants.HOST_LINKEDIN.getValue());
+        authorizeUriTemplate = LinkedinOAuth2Constants.AUTHORIZE_URI_TEMPLATE.getValue();
+        requiredScopes = LinkedinOAuth2Constants.REQUIRED_SCOPES.getValue();
+        scopeDelimiter = LinkedinOAuth2Constants.SCOPE_DELIMITER.getValue();
+        relayKey = LinkedinOAuth2Constants.RELAY_KEY.getValue();
+        authenticateUri = LinkedinOAuth2Constants.AUTHENTICATE_URI.getValue();
+        dataSource.addImportClass(DataSourceType.oauth2contact.name(), LinkedinContactsImport.class.getCanonicalName());
+    }
+
+    @Override
+    protected void validateTokenResponse(JsonNode response) throws ServiceException {
+        if (response.has(LinkedinErrorCodes.ERROR.getValue())) {
+            final String error = response.get(LinkedinErrorCodes.ERROR.getValue()).asText();
+            final JsonNode errorMsg = response.get(LinkedinErrorCodes.ERROR_DESCRIPTION.getValue());
+            ZimbraLog.extensions.debug("Response from linkedin: %s", response.asText());
+            switch (LinkedinErrorCodes.fromString(StringUtils.upperCase(error))) {
+            case USER_CANCELLED_LOGIN:
+                ZimbraLog.extensions.info(
+                    "User cancelled on login screen : " + errorMsg);
+                throw ServiceException.OPERATION_DENIED(
+                    "User cancelled on login screen");
+            case USER_CANCELLED_AUTHORIZE:
+                ZimbraLog.extensions.info(
+                        "User cancelled to authorize : " + errorMsg);
+                    throw ServiceException.OPERATION_DENIED(
+                        "User cancelled to authorize");
+            case DEFAULT_ERROR:
+            default:
+                ZimbraLog.extensions
+                    .warn("Unexpected error while trying to validate token: " + errorMsg);
+                throw ServiceException.PERM_DENIED("Token validation failed");
+            }
+        }
+
+        // ensure the tokens we requested are present
+        if (!response.has(LinkedinOAuth2Constants.ACCESS_TOKEN.getValue()) || !response.has(LinkedinOAuth2Constants.EXPIRES_IN.getValue())) {
+            throw ServiceException.PARSE_ERROR("Unexpected response from social service.", null);
+        }
+    }
+
+    @Override
+    protected String getPrimaryEmail(JsonNode credentials, Account account) throws ServiceException {
```

Comment:
Format this method's source (lines appear too long).

Reply:
formatted code